### DISCUSSION TITLE
GamePad Triggers (L2 and R2)

### DISCUSSION THREAD

**Discussion by luttje** [Key2Joy/Key2Joy#58] [🐕🦺 Tips & Tricks]
# 🎮 Tips on configuring L2 and R2

These buttons way down on the back of a GamePad are called 'Triggers'. 
<img src="https://github.com/luttje/Key2Joy/assets/2738114/0deb77a2-c62e-4bfe-91e1-6f5238bba443" width="256px" />

These are often referred to as the L2 and R2 buttons, but are quite different from buttons. Buttons can be pressed or not (binary), but the triggers can be pressed a bit (0), half way (.5), completely (1) and everything in between!

To react to a GamePad Trigger being pressed, you can add a trigger (confusingly named the same):
![279697743-48110ed7-d1eb-4750-a6e2-4311213d7ce9](https://github.com/luttje/Key2Joy/assets/2738114/44128f64-c2cb-4fef-aaf9-c594a1d1d14b)

Choose Left or Right to decide which trigger on the GamePad should activate this mapping.

## Mapping an action
There are several mappings you can make with a trigger input.

### You can [map a script action](https://github.com/luttje/Key2Joy/blob/main/Docs/Scripting.md#see-how-much-the-trigger-was-pressed) to this trigger. 
It could look something like this:
<img src="https://github.com/luttje/Key2Joy/raw/main/Docs/screenshot-scripting-trigger.png" width="512px" />

### You could map a GamePad Stick action:
This GamePad Stick action will use the amount the trigger is pressed to simulate moving the Right Stick on GamePad 1 to the right. If you set the `Trigger Input Delta Scale X` to `-1` it would move to the left:
<img src="https://github.com/luttje/Key2Joy/assets/2738114/3e4fdfe9-0a66-43bb-b2ce-fd3d5213824c" width="512px" />

When you now press your trigger on the gamepad, gamepad 1 will be simulated to have it's right stick move:
<img src="https://github.com/luttje/Key2Joy/assets/2738114/74e6f805-9435-4d38-bbb2-a13ded185b04" width="512px" />
   
As you can see, the stick stays to the right if we release the GamePad trigger quickly. This is because when the trigger returns to 0, it doesn't fire a trigger. Therefor we configure the stick to 0,0 after 500 ms. 

## 🐛 Notes
As you can see, the releasing of the trigger doesn't activate the mapping. This causes a GamePad Stick simulation to leave the stick hanging and requires a super janky workaround. I'll work on fixing this in the future (tracked in issue #59)

**Comment by bacmessi2001** [comment]
Hello, I'm having a problem due to my computer skills. I don't know how to simulate pressing the L2 and R2 keys completely (I want to assign them to a keyboard). Could you please give me specific instructions?

**Reply by luttje** [reply]
Hi! I hope this helps you:

**Install the latest version: https://github.com/Key2Joy/Key2Joy/releases/tag/0.8.0-alpha.1f3ef524cf:**
<img width="1047" height="523" alt="afbeelding" src="https://github.com/user-attachments/assets/225ee722-8129-4803-8048-89e77da8dca0" />

Follow the instructions to install it: right-click -> "Extract All" to unzip -> then start "Key2Joy.exe" -> click "More Info" -> then click "Run anyway"

After the app opens click the + button in the bottom right of the app:
<img width="1358" height="732" alt="afbeelding" src="https://github.com/user-attachments/assets/bdba8680-8909-4376-9468-f4b63b473d17" />

Then choose one of these options

### Option A: Doing it yourself, step-by-step

1. Setup the action as Keyboard Event and choose your key
    <img width="1224" height="499" alt="afbeelding" src="https://github.com/user-attachments/assets/c34106c7-d2af-488c-b801-45f90744dc43" />
2. Choose the GamePad Trigger Simulation with Left or Right trigger

3. Uncheck (make empty) Use input scale (delta from trigger input)

4. Make sure it says "Exact Delta: 1" (set it to 1, which means fully press down trigger)

5. Check (put the checkmark in it) the "Also create a child mapping that is a useful reverse of this"

6. Save Mapping

Follow the same steps for both left and right so you have both:
<img width="1409" height="441" alt="afbeelding" src="https://github.com/user-attachments/assets/3bad53e6-7287-4486-9019-6d12048647ea" />

### Option B: Loading my profile

If it's too many steps or too confusing, I made a profile for you with Q mapped to Left Trigger and E mapped to Right Trigger.

1. Download this file: [trigger-profile.k2j.json](https://github.com/user-attachments/files/27653740/trigger-profile.k2j.json) and place it somewhere on your PC. 

2. Then go to "File" -> "Load Profile"
    <img width="476" height="185" alt="afbeelding" src="https://github.com/user-attachments/assets/b5ac8e42-8f05-4dea-afee-9a89f8a68402" />

3. Load the file you just downloaded

It should work when you press Connect. Use Escape to stop.

**Reply by bacmessi2001** [reply]
Thank you very much, I have created a profile for myself
